### ISSUE TITLE
question about scaling -- chol.default positive definite error that is preventable by rescaling of trait variables

### ISSUE THREAD

**Issue by TeresaPegan** [hmsc-r/HMSC#173]
Hello! I have a question about scaling trait variables. I have a model where it seems that I need to rescale my trait variables before sampling to avoid an error. This was unexpected because I know HMSC rescales the variables automatically, so I am not sure why rescaling in advance would make a difference.

The model I am running includes 34 species, 1 variable in XData, and about 6 variables in TrData. The TrData variables come from a broad range of orders of magnitude (some on the order of 10, some on the order of 1e12, etc). In other linear modeling contexts I always rescale these predictors, but I did not do so with HMSC initially because I thought that HMSC will rescale them automatically.

However, my models with raw TrData variables consistently fail with the error
```
Error in chol.default(iV) : 
  the leading minor of order 2 is not positive definite
```
This error appeared consistently enough that I was never able to run a model past a couple thousand iterations before hitting it.
Based on reading other issues on this page (e.g. #45) and other internet sources, it seems that this error is difficult to debug, but I decided to try rescaling my variables. When I first tried this, I was not sure if I might be misunderstanding exactly what HMSC does to them by default.

It turns out that when I rescale my variables, I do not get this error. (At least I haven't yet). I rescaled all of my variables using the base R function scale(). Scale() sets the mean of the variable to about 0 and the sd to 1. After reading about it more, I think this is exactly the same as what HMSC is also supposed to be doing.

I was wondering if anyone has any insights into this issue! My thoughts are
1) I don't think there should be any problem if my data are rescaled twice (once by me in advance with scale(), once by HMSC internally). If the rescaling is happening in the same way twice, the second time around nothing will change. So I can just proceed with TrData variables that I rescale myself since it lets my models run without error.
2) However I am confused about why rescaling in advance would make a difference. Am I misunderstanding something about how HMSC rescales these variables? Is its rescaling method slightly different than that of scale()? Is this a sign that there is some other problem happening that I should dig further into?

It would be difficult for me to create a reproducible example for this error, but here are some more details about my model:

**model formula**
```
model1 =Hmsc(Y=Y_TP, XData=XData, XFormula=XFormula,TrData=TrData, 
             TrFormula=TrFormula, 
             phyloTree = phy, distr = "normal", 
             YScale = TRUE)
```
**Y data**
I am not using HMSC for species distribution data, but rather for another situation analogous to what is described in [Jiriniec et al 2021](https://www.science.org/doi/10.1126/sciadv.abk1743). I created my Y data as in this paper, which the authors describe in this quote --
"In our models, each individual bird was considered a sampling unit. Each row in our response variable input matrix thus represented an individual (denoted by band number), and each column represented 1 of our 77 species. For each row, the column corresponding to the correct species for that row had a numeric value (the response variable), and all other columns (species) in that row had a missing (NA) value."

**X data**
I have a single numeric variable in XData, which I have not rescaled. It ranges from 0 to 4 (it's not an integer, many of the values have decimals).

**Trait data**
I have 6 traits of interest which I have rescaled _a priori_ in order to get my models to run without chol positive definite errors. These 6 traits include 2 binary TRUE/FALSE traits and 4 continuous traits. The 4 continuous traits vary considerably in their order of magnitude, i.e. one of the traits ranges from 1e12 to 1e13, another ranges from 6 to 74, another ranges from 0 to 7000, etc).

**Phylogeny**
I have 34 species in this dataset and I have included a phylogenetic tree for them in the model.

**Comment by gtikhonov** [comment]
Hi! 

1. You can access the effective trait matrix by `model1$TrScaled` after running `model1 = Hmsc(...)`. Therefore, you may check whether automatic and manual rescaling indeed yield identical effective trait matrices.
2. With your trait values being that big, I suspect that there may be an overflow of capacity for floating point precision in involved matrix operations once we obtain `V` matrix for the original. This would especially be the likely case if you observe that the effective trait matrices are identical. Perhaps you may solve this issue by simply using trillions of units (10^12) for that trait instead of base units (10^0), so that the numbers are relatively small.

**Comment by TeresaPegan** [comment]
Hi,
1. That's great, thanks! I checked and the rescaled values are indeed identical whether I additionally rescale them in advance or not.
2. Ah, ok, if the original scales are involved in operations within the chains (i.e., the conversion back to the original scale is not something that happens only post-hoc, once the chain is finished), then I can easily see how having these huge-scale numbers would cause problems. Instead of rescaling prior to modeling, I converted my values to tractable units as you suggested, and everything has worked nicely. Thanks so much!
